### PR TITLE
Add options to triggers

### PR DESCRIPTION
**Pull request by betty-crokker**
This code satisfies both issues #166  and #167 

By having Permit() return a TriggerConfiguration instead of a StateConfiguration (where TriggerConfiguration is a subclass of StateConfiguration), we make it easy to add options to the trigger using fluent syntax.

For example:

```
sm.Configure(State.A)
    .Permit(Trigger.X, State.B).If(anonymousGuard, "description");

sm.Configure(State.A)
    .Permit(Trigger.X, State.B).If(() => true).Do(() => value = 1)
    .Permit(Trigger.X, State.C).If(() => false).Do(() => value = 2);

sm.Configure(State.A)
    .Permit(Trigger.X, State.B).If(() => true).Tag(1)
    .Permit(Trigger.X, State.C).If(() => false).Tag("blue");
```

### CODE REVIEW

File: src/Stateless/StateConfiguration.Async.cs [hunk 1]
```diff
@@ -24,8 +24,8 @@ public StateConfiguration InternalTransitionAsyncIf(TTrigger trigger, Func<bool>
             {
                 if (entryAction == null) throw new ArgumentNullException(nameof(entryAction));
 
-                _representation.AddTriggerBehaviour(new InternalTriggerBehaviour(trigger, guard));
-                _representation.AddInternalAction(trigger, (t, args) => entryAction(t));
+                Representation.AddTriggerBehaviour(new InternalTriggerBehaviour(trigger, guard));
```
Comment (HenningNT) [inline src/Stateless/StateConfiguration.Async.cs]: Did you make Representation public? If not, could we keep this as _representation?

**Comment by jsobell** [comment]
Nice. I've submitted a PR to your branch with the Tags being strongly typed too (but defaults to `object`)
Remains backwards compatible, but makes it a bit nicer than 'Object'.

File: src/Stateless/Transition.cs [hunk 2]
```diff
@@ -45,6 +48,11 @@ public Transition(TState source, TState destination, TTrigger trigger)
             public TTrigger Trigger { get { return _trigger; } }
 
             /// <summary>
+            /// Optional data associated with the transition.
+            /// </summary>
+            public object Tag { get { return _tag; } }
```
Comment (HenningNT) [inline src/Stateless/Transition.cs]: In my opinion the tag object is not really needed.

Comment (jsobell) [inline src/Stateless/Transition.cs]: I think it is, for reasons in a huge thread on the issues :)

Comment (jsobell) [inline src/Stateless/Transition.cs]: If you support guards, you need to be able to differentiate between transitions, and this is the only practical way we can come up with.
What I think is not necessary is the `Do()` feature, as that can be replaced with user specific implementation of the `Tag`, but I have no issue leaving it in as a helper feature.

Comment (HenningNT) [inline src/Stateless/Transition.cs]: I don't like to have one parameter do more than one thing, it reduces the clarity and makes it harder for a user to understand the API. So if we want to associate an Action with a trigger I prefer to have an explicit way of doing so. So let's keep Do(Action).

Comment (jsobell) [inline src/Stateless/Transition.cs]: No, neither do I. The 'Do()' is effectively redundant, as we already have an event that fires when the transition occurs, and the only thing missing is how to associate the two.
This is an incredibly common scenario, and has dozens of ways of coding. If you considered 'Tag' to be metadata then you might declare its type as `<int, Action<>>` and simply call the action in the central handler. 
Here's the OnOffExample modified to provide an ID and a function in the definition:
```
using System;
using Stateless;

namespace OnOffExample
{
    class TransitionMetadata
    {
        public int Id;
        public Action Activity;

        public TransitionMetadata(int id, Action activity)
        {
            this.Id = id;
            this.Activity = activity;
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                string on = "On", off = "Off";
                var space = ' ';

                var onOffSwitch = new StateMachine<string, char, TransitionMetadata>(off);
                onOffSwitch.OnTransitioned(t =>
                {
                    Console.WriteLine($"Detected transition ID {t.Tag.Id}");
                    t.Tag.Activity?.Invoke();
                });
                onOffSwitch.Configure(off).Permit(space, on).Tag(new TransitionMetadata(100, () => Console.WriteLine("On was transitioned!")));
                onOffSwitch.Configure(on).Permit(space, off).Tag(new TransitionMetadata(200, () => Console.WriteLine("Off was transitioned!")));

                Console.WriteLine("Press <space> to toggle the switch. Any other key will raise an error.");

                while (true)
                {
                    Console.WriteLine("Switch is in state: " + onOffSwitch.State);
                    var pressed = Console.ReadKey(true).KeyChar;
                    onOffSwitch.Fire(pressed);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Exception: " + ex.Message);
                Console.WriteLine("Press any key to continue...");
                Console.ReadKey(true);
            }
        }
    }
}
```

Sure, it's just another way of doing things, but it's a single point of execution for transitions, which I think is much cleaner. It also removes ambiguity between the event handler and the Do() action, so you define the order yourself.
Personally, I don't like code in my workflows. Ideally I like them to be serialisable, which means no embedded lambdas or object references, and when we are auto-generating workflows from users' definitions I would rather add those to a switch statement than embed them as lamdas throughout the StateMachine configuration section.

**Comment by betty-crokker** [comment]
Another benefit of having the options (e.g. "If" and "Do") in separate functions like that:

Right now we have 8 versions of OnEntryFrom(), representing (0 through 3 parameterized arguments) times (pass the transition as another argument yes or no).

If we follow that same pattern for PermitIf() and PermitDo() and PermitIfDo() we will have to add 6 versions of PermitIf, all 8 version of PermitDo, and 48 versions of PermitIfDo(), total of 62 new functions.

If instead we have Permit.If(), Permit.Do(), and Permit.If().Do(), we only have to write 14 new functions.

**Comment by betty-crokker** [comment]
Do we feel strongly about the version of PermitIf() that takes an array of Tuples?
1) It's an awkward syntax to use (all those "new Tuple<Func<bool, string>")
2) You have to pass a string description (or empty string or null which looks funny)
3) I don't see how it gets us anything beyond just calling all of the possible guard functions with logical OR between them:

```
sm.Configure(State.B)
    .PermitIf(Trigger.X, State.A,
        new Tuple<Func<bool>, string>(GreenLightIsOn, "GreenLightIsOn"),
        new Tuple<Func<bool>, string>(RedLightIsOff, "RedLightIsOff"));
```

could just be

```
sm.Configure(State.B)
    .Permit(Trigger.X, State.A).If(() => GreenLightIsOn() | RedLightIsOff());
```

**Comment by HenningNT** [comment]
One year later...
I think we should look into this again.
I would like to reduce the API, and the Permit.If().Do() syntax foes that. This would definitely be a breaking change, so it would go into stateless v5.
@betty-crokker would you like to update this PR sometime later, or maybe someone would like to do this?

**Comment by HenningNT** [comment]
Sorry, I'm closing this since it's quite old. Please let me know if you feel this is the wrong decision.

**Pull request closed without merging**